packages/hash/api/src/model/entityType.util.ts

```diff
@@ -0,0 +1,316 @@
+type PropertyGroup = {
+  parents: PropertyGroup[];
+  $id?: string;
+  properties: Property[];
+};
+
+type Property = {
+  name: string;
+  type: string;
+  format?: string;
+  description?: string;
+  otherFields: Partial<JSONSchema7>;
+};
+
+// If any property constraint has 'min' or 'max' in their name, it is a numeric constraint
+// JSON Schema 4 would allow booleans for 'exclusiveMinimum' and 'exclusiveMaximum', though.
+const CONSTRAINT_RE = /.*(min|max).*/i;
+
+/**
+ * When a 'minimim' constraint is defined twice on the same property (across multiple schemas),
+ * the constraint should be 'maximized' such that the minimum is changed to the most permissive
+ * value.
+ */
+const maximizeConstraint = Math.max;
+
+/**
+ * When a 'maximum' constraint is defined twice on the same property (across multiple schemas),
+ * the constraint should be 'minimized' such that the maximum is changed to the least permissive
+ * value.
+ */
+const minimizeConstraint = Math.min;
+
+/**
+ * All JSON Schema numeric constraints that we support validation for.
+ */
+type Constraint =
+  | "exclusiveMaximum"
+  | "exclusiveMinimum"
+  | "maximum"
+  | "minimum"
+  | "maxItems"
+  | "minItems"
+  | "maxLength"
+  | "minLength"
+  | "maxProperties"
+  | "minProperties";
+
+/**
+ * Mapping from constraint names to what kind of permissiveness that is required
+ */
+const propertyConstraintMerging: Record<
+  Constraint | string,
+  (...values: number[]) => number
+> = {
+  exclusiveMaximum: minimizeConstraint,
+  exclusiveMinimum: maximizeConstraint,
+  maximum: minimizeConstraint,
+  minimum: maximizeConstraint,
+  maxItems: minimizeConstraint,
+  minItems: maximizeConstraint,
+  maxLength: minimizeConstraint,
+  minLength: maximizeConstraint,
+  maxProperties: minimizeConstraint,
+  minProperties: maximizeConstraint,
+} as const;
+
+/**
+ * Traverse properties of a schema recursively, visiting 'allOf' properties as well.
+ *
+ * @param schema JSON Schema that is to be traversed
+ * @param properties list of Properties from the resulting traversal
+ */
+function traverseProperties(schema: JSONSchema7, properties: Property[]) {
+  // Traverse top-level properties for a schema
+  for (const [field, value] of Object.entries(schema.properties ?? {})) {
+    if (typeof value === "object") {
+      const { type, format, description, ...otherFields } = value;
+      if (typeof type === "string") {
+        properties.push({
+          name: field,
+          type,
+          format,
+          description,
+          otherFields,
+        });
+      }
+    }
+  }
+
+  // For any 'allOf' entry, check if more properties are to be uncovered.
+  for (const allOfEntry of schema.allOf ?? []) {
+    // AllOf might have more properties to check!
+    if (typeof allOfEntry === "object" && "type" in allOfEntry) {
+      if (allOfEntry.type === "object") {
+        traverseProperties(allOfEntry, properties);
+      }
+    }
+  }
```

Comment:
I don't understand why we're doing this. I don't think we should ever be storing actual schemas in the `allOf` array, just references - so why are we looking for properties here?

Perhaps more strongly typing the JSON schema would help get us on the same page about what we are actually storing in them.

Reply:
Yes that is true for creating schemas through the schema editor. This was assuming that people could submit their own schemas which might contain actual schemas in `allOf`

Reply:
Understood, that makes sense. My instinct is this is not going to be something we want to support (correct me if I'm wrong @CiaranMn ).

Instead of storing a json schema directly in an `allOf` array, we should enforce that schema creates create separate entity types and reference them in the array instead. Otherwise we will end up with un-captured entity types in our db, and a lot more considerations we would have to make in the schema editor UI.

If there is disagreement on this, perhaps the three of us should have an internal discussion about what the JSON schemas we are storing should look like so we can all get on the same page. My understanding was that the `allOf` array for our JSON schemas would only contain references to other schemas for now.

Reply:
I'm not sure what the use case of this is so I don't think we need to support it.

Reply:
In that case I suggest removing this for now. We can use the opportunity in the follow-up PR where we type our JSON schemas to assert that the `allOf` array can only contain references to other schemas, which will help disambiguate what our JSON schemas look like.

Perhaps it's worth also thinking about how we explain to block developers that our entity type JSON schemas are a subset of valid JSON Schemas.

Reply:
I, too, would prefer for `allOf` to only contain references, this was done to be defensive compared to what kind of schema we might receive. I would say we should keep it like this, adding a notice, and then removing it when we enforce `allOf` to only contain `$refs`. 

The reason being that some of the validation 'unit tests' (not proper unit tests) make use of the fact that we do not inherit from `$ref` but from schemas in `allOf`.

Reply:
Okay, I'll remove this then :)

Reply:
Thanks! Just want to avoid introducing validation logic for things we aren't supporting anyway. In future I think a well-designed `JSONSchema` TS type will prevent us from accounting for JSON-Schema features which we aren't supporting/making use of.